test/sig_utils.py

```diff
@@ -301,3 +312,365 @@ def reference_vector_argument(arg):
     if arg in ("array[] real", "row_vector"):
         return "vector"
     return arg
+
+overload_scalar = {
+    "Prim": "double",
+    "PrimExp": "double",
+    "Rev": "stan::math::var",
+    "RevSOA": "stan::math::var",
+    "RevExp": "stan::math::var",
+    "Fwd": "stan::math::fvar<double>",
+    "FwdExp": "stan::math::fvar<double>",
+    "Mix": "stan::math::fvar<stan::math::var>",
+}
+
+class CppStatement:
```

Comment:
It looks like you want to replace `stan::test::make_arg` with the "Argument" set of classes. So I think you should actually replace it.

Reply:
I wasn't thinking that but I'll take a look at it and see if it's worth doing. The goal here was making the code generation Python between varmat_compatibility_tests.py and generateExpressionsTests.py look the same.

Thinking positively I was hoping that the complexity was kinda net neutral on generateExpressionsTest.py. I guess you're right this new thing definitely isn't simpler on its own.

I think where I started on this was trying to work through the benchmarks scripts, where the logic on the expressions is much finickier ([here](https://github.com/stan-dev/math/blob/develop/benchmarks/benchmark.py#L347)). I had a version of the varmat tests that looked like that but I remember thinking it was hard to read ([here](https://github.com/stan-dev/math/blob/d71f157ad2c7b2de54130c7ba6d75298d9def0fe/test/test_varmat_signatures.py)).

The two things we'll get out of this though is that it's the same codegen used in varmat_compatibility_tests.py, and it's easy to test the codegen in the Python classes cause all this stuff is rather fragile (easy to write a test that always passes with codegen).

Anyway I'll investigate this more. If we truly decide this new stuff just sucks for the expressions tests and having the same kind of code gen isn't useful we can revert it.

Reply:
No, I am not trying to say that the abstractions you are adding are bad - just that they can be improved. More abstractions is almost always more code, but if it is not much more they can benefit readability. Also benchmarks could use these abstractions.

Reply:
Weeeee I changed this up a bit. Check out the generate expressions test code [now](https://github.com/stan-dev/math/blob/d1cab5237fdbe8214aa7573d53617558f11eac22/test/generateExpressionTests.py#L98).

I changed how the code runs to mimic the old expression tests, hid argument names, and added member functions to the FunctionGenerator for most of the utility functions you might call.

Now generateExpressionTests.py only needs to import and use the FunctionGenerator class, which is nice cause all the other little classes are hidden. I believe I will try to split the FunctionGenerator class into parsing-signatures and generating-code (cause these jobs are starting to look different).

```python
# Generate two sets of arguments, one will be expressions one will not
arg_list_expression_base = fg.build_arguments(len(fg.stan_args) * [overload], 2)
arg_list_expression = [fg.convert_to_expression(arg) if arg.is_matrix_like() else arg for arg in arg_list_expression_base]
arg_list_no_expression = fg.build_arguments(len(fg.stan_args) * [overload], 2)

# Check results with expressions and without are the same
cpp_function_name = f"stan::math::{fg.function_name}"
result = fg.function_call_assign(cpp_function_name, *arg_list_expression)
result_no_expression = fg.function_call_assign(cpp_function_name, *arg_list_no_expression)
fg.expect_eq(result, result_no_expression)

# Check that expressions evaluated only once
for arg in arg_list_expression:
  if arg.is_expression():
    fg.expect_leq_one(arg.counter)

# If reverse mode, check the adjoints            
if is_reverse_mode:
  summed_result = fg.recursive_sum(result)
  summed_result_no_expression = fg.recursive_sum(result_no_expression)
  sum_of_sums = fg.add(summed_result, summed_result_no_expression)
  fg.grad(sum_of_sums)

  for arg, arg_no_expression in zip(arg_list_no_expression, arg_list_expression):
    fg.expect_adj_eq(arg, arg_no_expression)

  fg.recover_memory()
```

And this generates the follow cpp:
```cpp
TEST(ExpressionTestRev, multiply_0) {
stan::math::var real0 = 0.4;
auto matrix1 = stan::test::make_arg<Eigen::Matrix<stan::math::var, 1, Eigen::Dynamic>>(0.4, 1);
int matrix1_expr2_counter = 0;
stan::test::counterOp<stan::math::var> matrix1_expr2_counter_op(&matrix1_expr2_counter);
auto matrix1_expr2 = matrix1.segment(0,1).unaryExpr(matrix1_expr2_counter_op);
stan::math::var real3 = 0.4;
auto matrix4 = stan::test::make_arg<Eigen::Matrix<stan::math::var, 1, Eigen::Dynamic>>(0.4, 1);
auto result5 = stan::math::eval(stan::math::multiply(real0,matrix1_expr2));
auto result6 = stan::math::eval(stan::math::multiply(real3,matrix4));
EXPECT_STAN_EQ(result5,result6);
EXPECT_LEQ_ONE(matrix1_expr2_counter);
auto summed_result7 = stan::math::eval(stan::test::recursive_sum(result5));
auto summed_result8 = stan::math::eval(stan::test::recursive_sum(result6));
auto sum_of_sums9 = stan::math::eval(stan::math::add(summed_result7,summed_result8));
stan::test::grad(sum_of_sums9);
stan::test::expect_adj_eq(real3,real0);
stan::test::expect_adj_eq(matrix4,matrix1_expr2);
stan::math::recover_memory();
}
```